### ISSUE TITLE
Documentation of events

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#2698]
## 📚 Documentation
Hi all, 

I just went through the documentation of events:
https://napari.org/guides/stable/connecting_events.html

There are some aspects I would love to read there as well:
* Events of the napari viewer, e.g. when it's closed.
* Event when the user navigates to another timepoint or z-slice
* Event when data in a layer was updated

If you tell me how to do this, I'll write the documentation/examples and send a PR.

Thanks!

Cheers,
Robert

**Comment by jni** [comment]
- when it's closed: don't think this exists as an event. Others might correct me here though.
- when the user navigates to another slice: `viewer.dims.events.current_step`, see example usage [here](https://github.com/napari/napari/blob/87a61fb420c7b9f479433f7e5e4bc2378817d17e/examples/points-over-time.py#L33-L35).
- when the data in a layer was updated: `layer.events.data`. This is *not* always reliable though: it is only emitted when the data is modified with `layer.data = whatever`, or through the viewer itself (because then we know we're changing it). It is *not* emitted e.g. when `.data` is a NumPy array and we do e.g. `viewer.add_image(my_image); my_image[5, 5] = 7`, because we have no way of knowing that happened. I'll also say that there's some places where `layer.events.data` is *not* emitted, e.g. when moving points. Those are bugs and there should be a `self.events.data()` call in those situations.

**Comment by tlambert03** [comment]
totally agree... the need for a table of available events has come up a number of times before.

> If you tell me how to do this, I'll write the documentation/examples and send a PR.

that would be amazing... and I can help.  Rather than focus on these three (useful, but somewhat arbitrary) events, I think we need a full table, like:

| class  | event | accessible at | description | emits | 
| ----- | ------ |---------------|------------|------|
| `Dims`  | current_step | `viewer.dims.events.current_step`  | emitted when the current dimension position changes | `(axis: int, raw_step: float)`

See https://github.com/napari/napari/issues/1389 where I started working on some code that would facilitate this by searching the codebase for all events, which could at least give us a table to start with, and then we can do a little manual curation (with the description part).  (also, ignore the "what it's hooked up to" part in that PR).

I'll see if I can get a start on this table soon, and it would be great if you wanted to help curate it

**Comment by goanpeca** [comment]
@tlambert03 would it help to try to self document to use `Event` when creating emitter groups and pass a `description="Some description"` to it?

Similar to how using pydantic we use `some_field : Type = Field(None, description="Some description", title="Some title")`

Seems like most of this could be fully automated as you stated in #1389

**Comment by tlambert03** [comment]
yeah, def.  That's a great idea

**Comment by tlambert03** [comment]
actually... the majority of our events _are_ generated from pydantic `Fields` ... so we can literally use that!

**Comment by tlambert03** [comment]
Ok, looking into this a bit more, I think maybe the best place to start _is_ just a markdown table that we can populate.  It will be the easiest way to have people help us document this.  Once that table is populated, we can put it in the source code as `Field` descriptions.  It's a little complicated at the moment, since some of our events are autogenerated by `EventedModel`, but `Layers` aren't yet converted over to the new model, and as such dynamically declare their fields in the `__init__`, so it's hard to parse them all using the same method.

@haesleinhuepf, here is a script (gist [here](https://gist.github.com/tlambert03/cdbed6bfa19e85efc869815cb760f319)) that will generate the following table...

<details>

<summary>script</summary>

```python
from dataclasses import dataclass
from typing import Optional, Type

import napari
import inspect
from napari.utils.events import EventedModel
from napari.utils.settings._defaults import BaseNapariSettings
from napari.components.viewer_model import ViewerModel


def walk_modules(module, pkg='napari', _walked=None):
    if not _walked:
        _walked = set()
    yield module
    _walked.add(module)
    for name in dir(module):
        attr = getattr(module, name)
        if (
            inspect.ismodule(attr)
            and attr.__package__.startswith(pkg)
            and attr not in _walked
        ):
            yield from walk_modules(attr, pkg, _walked=_walked)


def iter_classes(module):
    for name in dir(module):
        attr = getattr(module, name)
        if inspect.isclass(attr) and attr.__module__ == module.__name__:
            yield attr


@dataclass
class Ev:
    name: str
    model: Type
    description: str = ''
    type_: Optional[Type] = None

    def on_viewer(self):
        if issubclass(self.model, ViewerModel):
            return f'viewer.events.{self.name}'
        for name, field_ in napari.Viewer.__fields__.items():
            if field_.type_ is self.model:
                return f'viewer.{name}.events.{self.name}'
        return ''

    def type_name(self):
        if cls_name := getattr(self.type_, '__name__', None):
            return cls_name
        name = str(self.type_) if self.type_ else ''
        return name.replace("typing.", "")


HEADER = ('Class', 'Event Name', 'From viewer', 'Description', 'Event Attributes')
ROW = '| {:13.13} | {:16.16} | {:37.37} | {:20.20} | {:37.37} |'


def iter_events(module=napari):
    for mod in walk_modules(module):
        for kls in iter_classes(mod):
            if not issubclass(kls, EventedModel) or issubclass(
                kls, BaseNapariSettings
            ):
                continue
            for name, field_ in kls.__fields__.items():
                finfo = field_.field_info
                if finfo.allow_mutation:
                    yield Ev(name, kls, finfo.description, field_.type_)


if __name__ == '__main__':
    print(ROW.format(*HEADER))
    print(ROW.format(*('-' * 60,) * 5))

    for ev in iter_events():
        if ev.on_viewer():
            row = ROW.format(
                ev.model.__name__,
                ev.name,
                ev.on_viewer(),
                ev.description or '',
                f'value: {ev.type_name()}',
            )
            print(row)
```

</details>

### table:
| Class         | Event Name       | From viewer                              | Description          | Event Attributes                      |
| ------------- | ---------------- | ---------------------------------------- | -------------------- | ------------------------------------- |
| `Axes`        | visible          | `viewer.axes.events.visible`             |                      | value: `bool`                         |
| `Axes`        | labels           | `viewer.axes.events.labels`              |                      | value: `bool`                         |
| `Axes`        | colored          | `viewer.axes.events.colored`             |                      | value: `bool`                         |
| `Axes`        | dashed           | `viewer.axes.events.dashed`              |                      | value: `bool`                         |
| `Axes`        | arrows           | `viewer.axes.events.arrows`              |                      | value: `bool`                         |
| `Camera`      | center           | `viewer.camera.events.center`            |                      | value: `Tuple[float, float, float]`   |
| `Camera`      | zoom             | `viewer.camera.events.zoom`              |                      | value: `float`                        |
| `Camera`      | angles           | `viewer.camera.events.angles`            |                      | value: `Tuple[float, float, float]`   |
| `Camera`      | perspective      | `viewer.camera.events.perspective`       |                      | value: `float`                        |
| `Camera`      | interactive      | `viewer.camera.events.interactive`       |                      | value: `bool`                         |
| `Cursor`      | position         | `viewer.cursor.events.position`          |                      | value: `float`                        |
| `Cursor`      | scaled           | `viewer.cursor.events.scaled`            |                      | value: `bool`                         |
| `Cursor`      | size             | `viewer.cursor.events.size`              |                      | value: `int`                          |
| `Cursor`      | style            | `viewer.cursor.events.style`             |                      | value: `CursorStyle`                  |
| `Dims`        | ndim             | `viewer.dims.events.ndim`                |                      | value: `int`                          |
| `Dims`        | ndisplay         | `viewer.dims.events.ndisplay`            |                      | value: `Literal[2, 3]`                |
| `Dims`        | last_used        | `viewer.dims.events.last_used`           |                      | value: `int`                          |
| `Dims`        | range            | `viewer.dims.events.range`               |                      | value: `Tuple[float, float, float]`   |
| `Dims`        | current_step     | `viewer.dims.events.current_step`        |                      | value: `int`                          |
| `Dims`        | order            | `viewer.dims.events.order`               |                      | value: `int`                          |
| `Dims`        | axis_labels      | `viewer.dims.events.axis_labels`         |                      | value: `str`                          |
| `GridCanvas`  | enabled          | `viewer.grid.events.enabled`             |                      | value: `bool`                         |
| `GridCanvas`  | stride           | `viewer.grid.events.stride`              |                      | value: `int`                          |
| `GridCanvas`  | shape            | `viewer.grid.events.shape`               |                      | value: `Tuple[int, int]`              |
| `ScaleBar`    | visible          | `viewer.scale_bar.events.visible`        |                      | value: `bool`                         |
| `ScaleBar`    | colored          | `viewer.scale_bar.events.colored`        |                      | value: `bool`                         |
| `ScaleBar`    | ticks            | `viewer.scale_bar.events.ticks`          |                      | value: `bool`                         |
| `ScaleBar`    | position         | `viewer.scale_bar.events.position`       |                      | value: `Position`                     |
| `ScaleBar`    | font_size        | `viewer.scale_bar.events.font_size`      |                      | value: `float`                        |
| `ScaleBar`    | unit             | `viewer.scale_bar.events.unit`           |                      | value: `str`                          |
| `TextOverlay` | visible          | `viewer.text_overlay.events.visible`     |                      | value: `bool`                         |
| `TextOverlay` | color            | `viewer.text_overlay.events.color`       |                      | value: `Array`                        |
| `TextOverlay` | font_size        | `viewer.text_overlay.events.font_size`   |                      | value: `float`                        |
| `TextOverlay` | position         | `viewer.text_overlay.events.position`    |                      | value: `TextOverlayPosition`          |
| `TextOverlay` | text             | `viewer.text_overlay.events.text`        |                      | value: `str`                          |
| `ViewerModel` | help             | `viewer.events.help`                     |                      | value: `str`                          |
| `ViewerModel` | status           | `viewer.events.status`                   |                      | value: `str`                          |
| `ViewerModel` | theme            | `viewer.events.theme`                    |                      | value: `str`                          |
| `ViewerModel` | title            | `viewer.events.title`                    |                      | value: `str`                          |
| `Viewer`      | help             | `viewer.events.help`                     |                      | value: `str`                          |
| `Viewer`      | status           | `viewer.events.status`                   |                      | value: `str`                          |
| `Viewer`      | theme            | `viewer.events.theme`                    |                      | value: `str`                          |
| `Viewer`      | title            | `viewer.events.title`                    |                      | value: `str`                          |


so if we started writing down simple descriptions for each one, that would be a great thing to put in the events documentation, and eventually add to the source code as `description` fields

**Comment by GenevieveBuckley** [comment]
This is great @tlambert03

**Comment by GenevieveBuckley** [comment]
Should this issue have been closed by https://github.com/napari/napari/pull/2750?

We should either close the issue, or describe what is left remaining to do here.

**Comment by tlambert03** [comment]
Thanks @GenevieveBuckley 
Yes #2750 was addressing this.  I'd say close this and open a new, more focused issue with any additional proposals for improving our event docs
